torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -460,44 +452,229 @@ void CompiledFusionFunction::launch_with_tensors(at::ArrayRef<at::Tensor> inputs
-  int numBlocks = std::min(maxBlocks, ceilDiv(numel, blockSize));
-  //std::cout << "maxBlocks = " << maxBlocks << " needed blocks: " << ceilDiv(numel,blockSize)
-  //          << " numblocks =  " << numBlocks;
-
-  // it is possible that this is the first cuda call on this thread
-  // so make sure we initialize the Driver API's context
-  // cudaFree(0) accomplishes this.
-  cudaFree(0);
-
-  TORCH_CU_CHECK(cuLaunchKernel(
-    function,
-    numBlocks, 1, 1,
-    blockSize, 1, 1,
-    0, nullptr,
-    arguments,
-    nullptr));
+#ifdef WITH_CUDA
+
+struct CUDAFusionFunction : public CompiledFusionFunction {
+  CUDAFusionFunction(const std::string & name, AnnotatedGraph & agraph)
+  : CompiledFusionFunction(name, agraph) {
+    TORCH_CUDA_CHECK(cudaGetDevice(&device));
+    TORCH_CUDA_CHECK(cudaGetDeviceProperties(&prop, device));
+
+    if ((prop.major >= 6 && CUDA_VERSION < 8000) ||
+        (prop.major >= 7 && CUDA_VERSION < 9000)) {
+      std::stringstream err_string;
+      err_string << "In CompiledFusionFunction, PyTorch compiled with insufficient CUDA version: "
+  	       << CUDA_VERSION << " for the current GPU device " << prop.name
+  	       << " with device capability " << prop.major << "." << prop.minor;
+      throw std::runtime_error(err_string.str());
+    }
+
+    std::stringstream cu;
+    concat_desc = codegen::emitCompilationUnit(cu, name, agraph, true);
+    compilation_unit = cu.str();
+    nvrtcProgram program;
+    TORCH_NVRTC_CHECK(nvrtcCreateProgram(&program, compilation_unit.c_str(), NULL, 0, nullptr, nullptr));
+
+    std::string compute = "--gpu-architecture=compute_" + std::to_string(prop.major) + std::to_string(prop.minor);
+    std::vector<const char *> args = {"--std=c++11", compute.c_str()};
+    nvrtcResult result = nvrtcCompileProgram(program, args.size(), args.data());
+    if (result == NVRTC_ERROR_COMPILATION) {
+      size_t logsize;
+      nvrtcGetProgramLogSize(program, &logsize);
+      std::vector<char> log(logsize);
+      nvrtcGetProgramLog(program, log.data());
+      cu << log.data();
+      throw std::runtime_error(cu.str());
+    }
+    ResourceGuard holdProgram([&] {
+      TORCH_NVRTC_CHECK(nvrtcDestroyProgram(&program));
+    });
+    TORCH_NVRTC_CHECK(result);
+
+    size_t ptx_size;
+    TORCH_NVRTC_CHECK(nvrtcGetPTXSize(program, &ptx_size));
+    ptx.resize(ptx_size);
+    TORCH_NVRTC_CHECK(nvrtcGetPTX(program, ptx.data()));
+
+    TORCH_CU_CHECK(cuModuleLoadData(&module, ptx.data()));
+    TORCH_CU_CHECK(cuModuleGetFunction(&function, module, name.c_str()));
+
+    TORCH_CU_CHECK(cuOccupancyMaxActiveBlocksPerMultiprocessor(
+      &maxBlocks, function, 128, 0));
+    maxBlocks *= prop.multiProcessorCount;
+  }
+  virtual ~CUDAFusionFunction() override {
+    TORCH_CU_CHECK(cuModuleUnload(module));
+  }
+protected:
+  virtual at::Backend backend() const override {
+    return at::kCUDA;
+  }
+  virtual void launch_raw(uint32_t numel, void ** arguments) override {
```

Comment:
Can you add a comment describing what `arguments` are exactly (probably in the header_?